Tests/ApolloCacheDependentTests/LoadQueryFromStoreTests.swift

```diff
@@ -283,80 +227,58 @@ class LoadQueryFromStoreTests: XCTestCase, CacheTesting {
     let starshipLength = 1234.5
     let coordinates = [[38.857150, -94.798464]]
-
-    let initialRecords: RecordSet = [
+    
+    mergeRecordsIntoCache([
       "QUERY_ROOT": ["starshipCoordinates(coordinates:\(coordinates))": Reference(key: "starshipCoordinates(coordinates:\(coordinates))")],
       "starshipCoordinates(coordinates:\(coordinates))": ["__typename": "Starship",
                                                           "name": "Millennium Falcon",
                                                           "length": starshipLength,
                                                           "coordinates": coordinates]
-    ]
-
-    withCache(initialRecords: initialRecords) { (cache) in
-      store = ApolloStore(cache: cache)
-
-      let query = StarshipCoordinatesQuery(coordinates: coordinates)
-
-      load(query: query) { result in
-        switch result {
-        case .success(let graphQLResult):
-          XCTAssertNil(graphQLResult.errors)
-
-          guard let data = graphQLResult.data else {
-            XCTFail("No data returned with result")
-            return
-          }
-
-          XCTAssertEqual(data.starshipCoordinates?.name, "Millennium Falcon")
-          XCTAssertEqual(data.starshipCoordinates?.length, starshipLength)
-          XCTAssertEqual(data.starshipCoordinates?.coordinates, coordinates)
-        case .failure(let error):
-          XCTFail("Unexpected error: \(error)")
-        }
-      }
+    ])
+    
+    let query = StarshipCoordinatesQuery(coordinates: coordinates)
+    
+    loadFromStore(query: query) { result in
+      let graphQLResult = try result.get()
+      XCTAssertEqual(graphQLResult.source, .cache)
+      XCTAssertNil(graphQLResult.errors)
+      
+      let data = try XCTUnwrap(graphQLResult.data)
+      XCTAssertEqual(data.starshipCoordinates?.name, "Millennium Falcon")
+      XCTAssertEqual(data.starshipCoordinates?.length, starshipLength)
+      XCTAssertEqual(data.starshipCoordinates?.coordinates, coordinates)
     }
   }
 
   // MARK: - Helpers
   
-  private func load<Query: GraphQLQuery>(query: Query, resultHandler: @escaping GraphQLResultHandler<Query.Data>) {
-    let expectation = self.expectation(description: "Loading query from store")
+  private func loadFromStore<Query: GraphQLQuery>(query: Query, file: StaticString = #filePath, line: UInt = #line, resultHandler: @escaping AsyncResultObserver<GraphQLResult<Query.Data>, Error>.ResultHandler) {
+    let resultObserver = makeResultObserver(for: query, file: file, line: line)
+        
+    let expectation = resultObserver.expectation(description: "Loaded query from store", file: file, line: line, resultHandler: resultHandler)
     
-    store.load(query: query) { result in
-      resultHandler(result)
-      expectation.fulfill()
-    }
+    store.load(query: query, resultHandler: resultObserver.handler)
     
-    waitForExpectations(timeout: 5, handler: nil)
+    wait(for: [expectation], timeout: defaultWaitTimeout)
```

Comment:
Yeah, the `resultObserver` is responsible for fulfilling the expectation. That makes the most sense in situations where you set up multiple expectations, and it would be more confusing to keep track of which expectation belongs to which handler (or at least that was my thinking).

Here, the only reason this is using a result observer instead of a stand-alone expectation is that the handler can throw because we're using `result.get()` and `XCTUnwrap`. We can get rid of `result.get()` by switching to `XCTAssertSuccessResult`, but I would really like to keep using `XCTUnwrap` (or come up with an alternative, but not sure what that could look like).

Reply:
I feel like a recurring theme of our conversations is "This thing that makes sense for complex things seems really weird for less complex things" 🙃

I'm definitely fine with using `XCTUnwrap`, and this definitely reads better at the point it's called then having all this gunk in every test. I think I just need to get my head around some of these helpers that you're writing. It'll probably help for me to write some tests with them after this gets merged.